DataFormats/Detectors/TRD/include/DataFormatsTRD/TrapConfig3.h

```diff
@@ -0,0 +1,619 @@
+
+  // get the indenfiying characteristics of a register given its address.
+  void getRegisterByAddr(uint32_t registeraddr, std::string& regname, int32_t& newregidx, int32_t& numberbits);
+
+  // return all the registers for a particular mcm
+  void getAllRegisters(int mcmidx, std::array<uint32_t, kLastReg>& registers);
+
+  // return all the mcm values for a particular register index
+  void getAllMCMByIndex(int regindex, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& registers);
+
+  // return all the mcm values for a particular register name
+  void getAllMCMByName(std::string registername, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& mcms);
+
+  // return all the mcm for a particular register address
+  void getAllMCMByAddress(int registeraddress, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& mcms);
+
+  // return all the config data in an unpacked array.
+  void getAll(std::array<uint32_t, kLastReg * o2::trd::constants::MAXMCMCOUNT>& configdata);
+
+  // return the full unpacked config event.
+  bool printRegister(int addr, int det, int rob, int mcm);
+  bool printRegister(TrapRegInfo* reg, int det, int rob, int mcm);
+
+  const std::array<o2::trd::TrapRegInfo, kLastReg>& getTrapRegisters() { return mTrapRegisters; }
+
+  // population pending
+  void setConfigVersion(uint32_t version) { mTrapConfigVersion = version; }           // these must some how be gotten from git or wingdb.
+  void setConfigNumber(uint32_t number) { mTrapConfigNumber = number; }               // these must be gotten from git or wingdb.
+  void setConfigSavedVersion(uint16_t version) { mTrapConfigSavedVersion = version; } // the version that is saved, for the ability to later save the config differently.
+  uint32_t getConfigVersion() { return mTrapConfigVersion; }                          // these must some how be gotten from git or wingdb.
+  uint32_t getConfigName() { return mTrapConfigNumber; }                              // these must be gotten from git or wingdb.
+  uint16_t getConfigSavedVersion() { return mTrapConfigSavedVersion; }                // the version that is saved, for the ability to later save the config differently.
+
+  inline bool operator==(const TrapConfig3& other);
+
+  // for compliance with the same interface to o2::trd::TrapConfig and its run1/2 inherited interface:
+  // only these 2 are used in the simulations.
+  uint32_t getDmemUnsigned(uint32_t address, int detector, int rob, int mcm);
+  uint32_t getTrapReg(uint32_t index, int detector, int rob, int mcm);
+
+ private:
+  std::array<o2::trd::TrapRegInfo, kLastReg> mTrapRegisters;                              //!< store of layout of each block of mTrapRegisterSize, populated via initialiseRegisters
+  std::bitset<o2::trd::constants::MAXMCMCOUNT> mMCMDataHasChanged;                        //!< does the mcm actually contain data.
+  std::array<uint32_t, kTrapRegistersSize * o2::trd::constants::MAXMCMCOUNT> mConfigData; //!< one block of data per mcm.
```

Comment:
Wouldn't an array of arrays make the indexing simpler.
Also note that as discussed elsewhere, this seems to reserve data for non-existent MCMs (C0 chambers in stack 2 with only 6 ROBs). It also seems that BM/HCM MCMs are not stored at all.

Reply:
BM/HCM comment above. Yes array of array would be simpler, and was the reason I asked about this in another PR a while back, have not changed it yet.